libraries/abstractions/transport/secure_sockets/transport_secure_sockets.c

```diff
@@ -0,0 +1,511 @@
+ * @param[in] pServerInfo Server connection info.
+ * @param[in] pSocketsConfig socket configs for the connection.
+ *
+ * @return #TRANSPORT_SOCKET_STATUS_SUCCESS on success;
+ *         #TRANSPORT_SOCKET_STATUS_INVALID_PARAMETER, #TRANSPORT_SOCKET_STATUS_INSUFFICIENT_MEMORY,
+ *         #TRANSPORT_SOCKET_STATUS_INVALID_CREDENTIALS, #TRANSPORT_SOCKET_STATUS_INTERNAL_ERROR,
+ *         #TRANSPORT_SOCKET_STATUS_DNS_FAILURE, #TRANSPORT_SOCKET_STATUS_CONNECT_FAILURE on failure.
+ */
+static TransportSocketStatus_t establishConnect( NetworkContext_t * pNetworkContext,
+                                                 const ServerInfo_t * pServerInfo,
+                                                 const SocketsConfig_t * pSocketsConfig );
+
+/**
+ * @brief Set up TLS configurations for the socket.
+ *
+ * @param[in] pSocketsConfig The session configuration.
+ * @param[in] tcpSocket An initialized socket to secure.
+ * @param[in] pHostName Remote server name for SNI.
+ * @param[in] hostnameLength The length of `pHostName`.
+ *
+ * @return #SOCKETS_ERROR_NONE or #SOCKETS_SOCKET_ERROR.
+ */
+static int32_t tlsSetup( const SocketsConfig_t * pSocketsConfig,
+                         Socket_t tcpSocket,
+                         const char * pHostName,
+                         size_t hostnameLength );
+
+
+/*-----------------------------------------------------------*/
+
+int32_t SecureSocketsTransport_Send( const NetworkContext_t * pNetworkContext,
+                                     const void * pMessage,
+                                     size_t bytesToSend )
+{
+    int32_t bytesSent = 0;
+    int32_t transportSocketStatus = SOCKETS_ERROR_NONE;
+
+    if( ( pMessage == NULL ) ||
+        ( bytesToSend <= 0UL ) ||
+        ( pNetworkContext == NULL ) )
+    {
+        LogError( ( "Invalid parameter: pMessage=%p, bytesToSend=%lu, pNetworkContext=%p",
+                    pMessage, bytesToSend, pNetworkContext ) );
+        bytesSent = SOCKETS_EINVAL;
+    }
+    else if( ( pNetworkContext != NULL ) && ( pNetworkContext->pContext == NULL ) )
+    {
+        LogError( ( "Invalid parameter: pNetworkContext->pContext cannot be NULL." ) );
+        bytesSent = SOCKETS_EINVAL;
+    }
+    else
+    {
+        transportSocketStatus = SOCKETS_Send( ( Socket_t ) pNetworkContext->pContext,
+                                              pMessage,
+                                              bytesToSend,
+                                              0 );
+
+        /* If an error occurred, a negative value is returned. @ref SocketsErrors. */
+        if( transportSocketStatus >= 0 )
+        {
+            bytesSent = transportSocketStatus;
+            LogInfo( ( "Successfully sent %ld bytes over network.", transportSocketStatus ) );
+        }
+        else
+        {
+            LogError( ( "Failed to send data over network. BytesRequested=%ld.", transportSocketStatus ) );
+            bytesSent = SOCKETS_SOCKET_ERROR;
```

Comment:
How come we don't return transportSocketStatus on failure?

Reply:
this API returns how much bytes it received, bytesSent is the negative value when we encounter error.

It is similar to the [CSDK's API](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/master/platform/posix/transport/include/openssl_posix.h#L178)

Reply:
simplified the logic to avoid confusion in 3e78d0f4ec89db890c2801c1d13e61bceb12f869